sdks/java/io/hadoop-output-format/src/main/java/org/apache/beam/sdk/io/hadoop/outputformat/HadoopOutputFormatIO.java

```
@@ -196,97 +211,343 @@ public void populateDisplayData(DisplayData.Builder builder) {
         builder.addIfNotNull(
-            DisplayData.item(OUTPUTFORMAT_VALUE_CLASS, hadoopConfig.get(OUTPUTFORMAT_VALUE_CLASS))
+            DisplayData.item(
+                    OUTPUT_FORMAT_VALUE_CLASS_ATTR,
+                    hadoopConfig.get(OUTPUT_FORMAT_VALUE_CLASS_ATTR))
                 .withLabel("OutputFormat Value Class"));
+        builder.addIfNotNull(
+            DisplayData.item(PARTITIONER_CLASS_ATTR, hadoopConfig.get(PARTITIONER_CLASS_ATTR))
+                .withLabel("Partitioner Class"));
+        builder.addIfNotNull(
+            DisplayData.item(NUM_REDUCES, hadoopConfig.get(NUM_REDUCES))
+                .withLabel("Reducers Count"));
+      }
+    }
+
+    private void setupJob(SerializableConfiguration configuration) {
+      try {
+        JobID jobId = HadoopUtils.createJobId();
+        TaskAttemptContext setupTaskContext =
+            HadoopUtils.createSetupTaskContext(getConfiguration().get(), jobId);
+        OutputFormat<?, ?> jobOutputFormat =
+            HadoopUtils.createOutputFormatFromConfig(getConfiguration().get());
+        jobOutputFormat.checkOutputSpecs(setupTaskContext);
+        jobOutputFormat.getOutputCommitter(setupTaskContext).setupJob(setupTaskContext);
+
+        configuration.get().set(MRJobConfig.ID, jobId.getJtIdentifier());
+      } catch (Exception e) {
+        throw new RuntimeException("Unable to setup job.", e);
       }
     }
 
     @Override
-    public PDone expand(PCollection<KV<K, V>> input) {
-      input.apply(ParDo.of(new WriteFn<>(this)));
-      return PDone.in(input.getPipeline());
+    public PDone expand(PCollection<KV<KeyT, ValueT>> input) {
+
+      setupJob(getConfiguration());
+
+      // TODO add branch for streaming
+      PCollectionView<SerializableConfiguration> configView = null;
+      if (input.getWindowingStrategy().equals(WindowingStrategy.globalDefault())) {
+        configView = createGlobalConfigCollectionView(input);
+      }
```

we should throw an exception for non default windowing

---

I will fix it in the streaming iteration